### PR TITLE
Add Symfony 3 support

### PR DESCRIPTION
**Pull request by WouterJ**
Replaces https://github.com/schmittjoh/JMSSecurityExtraBundle/pull/187

Changes done on top of that PR:

 * Inject security.context in older versions
 * Use better composer.json version constraints
 * Update Travis configuration and test matrix

### CODE REVIEW

File: Resources/config/security_expressions.xml [hunk 1]
```diff
@@ -27,7 +27,7 @@
 
     <services>
         <service id="security.extra.twig_extension" class="%security.extra.twig_extension.class%" public="false">
-            <argument type="service" id="security.context" />
+            <argument type="service" id="security.authorization_checker" />
```
Comment (Ener-Getick) [inline Resources/config/security_expressions.xml]: Isn't it safer to add ``on-invalid="ignore"`` in the case the core validation pass is executed before ``SecurityCompatibilityPass`` ?

File: composer.json [hunk 2]
```diff
@@ -11,34 +11,32 @@
...
-        "symfony/browser-kit": "*",
-        "symfony/finder": "*",
-        "symfony/css-selector": "*",
-        "symfony/process": "*",
-        "doctrine/doctrine-bundle": "*",
-        "symfony/twig-bundle": "*",
-        "doctrine/orm": "*",
-        "symfony/form": "*",
-        "symfony/validator": "*"
+        "sensio/framework-extra-bundle": "~2.0|~3.0",
```
Comment (Ener-Getick) [inline composer.json]: IMO you can update this constraints and only supports ``sf >= 2.3``.

**Comment by Ener-Getick** [comment]
Except my comments :+1:

**Comment by Ener-Getick** [comment]
The tests have to be fixed too.

File: Resources/config/security_expressions.xml (hunk 1 above)
Comment (Ener-Getick) [inline Resources/config/security_expressions.xml]: Ok there are always executed after the normal compilation passes.

**Comment by Jean85** [comment]
I agree with @Ener-Getick

**Comment by Ninir** [comment]
This rocks. Can't wait to have it release! good job @WouterJ

**Comment by Ener-Getick** [comment]
@WouterJ do you need help to fix the tests?

**Comment by WouterJ** [comment]
@Ener-Getick I'm working on it now. I think I can manage to fix them

**Comment by Ener-Getick** [comment]
Ok cool :-)

File: Twig/SecurityExtension.php [hunk 3]
```diff
@@ -3,27 +3,33 @@
 namespace JMS\SecurityExtraBundle\Twig;
 
 use JMS\SecurityExtraBundle\Security\Authorization\Expression\Expression;
+use JMS\SecurityExtraBundle\Exception\InvalidArgumentException;
+use Symfony\Component\Security\Core\Authorization\AuthorizationChecker;
```
Comment (Ener-Getick) [inline Twig/SecurityExtension.php]: Isn't it safer to use ``AuthorizationCheckerInterface``?

**Comment by Ener-Getick** [comment]
You should also use ``Psr\Log\NullLogger`` [here](https://github.com/schmittjoh/JMSSecurityExtraBundle/blob/master/Tests/Security/Util/SecureRandomTest.php#L6).

**Comment by Ener-Getick** [comment]
You also have to use ``Psr\Log\LoggerInterface`` in many places:
* [``JMS\SecurityExtraBundle\Security\Acl\Expression\PermissionEvaluator``](https://github.com/schmittjoh/JMSSecurityExtraBundle/blob/master/Security/Acl/Expression/PermissionEvaluator.php#L28)
* [``JMS\SecurityExtraBundle\Security\Acl\Voter\AclVoter``](https://github.com/schmittjoh/JMSSecurityExtraBundle/blob/master/Security/Acl/Voter/AclVoter.php#L6)
*  [``JMS\SecurityExtraBundle\Security\Authorization\AfterInvocation\AclAfterInvocationProvider``](https://github.com/schmittjoh/JMSSecurityExtraBundle/blob/master/Security/Authorization/AfterInvocation/AclAfterInvocationProvider.php#L21)
* [``JMS\SecurityExtraBundle\Security\Authorization\Expression\ExpressionVoter``](https://github.com/schmittjoh/JMSSecurityExtraBundle/blob/master/Security/Authorization/Expression/ExpressionVoter.php#L21)
* [``JMS\SecurityExtraBundle\Security\Authorization\Interception\MethodSecurityInterceptor``](https://github.com/schmittjoh/JMSSecurityExtraBundle/blob/master/Security/Authorization/Interception/MethodSecurityInterceptor.php#L30)
* [``JMS\SecurityExtraBundle\Security\Util\SecureRandom``](https://github.com/schmittjoh/JMSSecurityExtraBundle/blob/master/Security/Util/SecureRandom.php#L7)

File: .travis.yml [hunk 4]
```diff
@@ -1,15 +1,37 @@
 language: php
 
+sudo: false
+
+cache:
+  directories:
+    - "$HOME/.composer/cache"
```
Comment (stof) [inline .travis.yml]: I suggest caching only the ``files`` subdolder. See https://github.com/puli/manager/pull/9#discussion_r40499480 for the explanation

File: .travis.yml [hunk 5]
```diff
@@ -1,15 +1,37 @@
...
+    - php: 5.6
+      env: SYMFONY_VERSION="2.7.*"
+    - php: 5.6
+      env: SYMFONY_VERSION="2.8.*"
+    - php: 5.6
+      env: SYMFONY_VERSION="3.0.*"
+    - php: 5.6
+      env: SYMFONY_VERSION="3.1.*"
+  allowed_failure:
+    - env: SYMFONY_VERSION="3.1.*"
```
Comment (stof) [inline .travis.yml]: this one should be better handled with a job testing against dev versions of dependencies

File: Twig/SecurityExtension.php [hunk 6]
```diff
@@ -3,27 +3,33 @@
...
 class SecurityExtension extends \Twig_Extension
 {
-    private $context;
+    private $authorizationChecker;
 
-    public function __construct(SecurityContextInterface $context)
+    public function __construct($authorizationChecker)
     {
-        $this->context = $context;
+        if (!$authorizationChecker instanceof SecurityContextInterface && !$authorizationChecker instanceof AuthorizationChecker) {
```
Comment (stof) [inline Twig/SecurityExtension.php]: the check must use the interface, to allow composition

**Comment by Jean85** [comment]
Travis is failing on PHP 5.3 for too much memory usage. You should disable xDebug (at least) there.
To do that on Travis, you should add `phpenv config-rm xdebug.ini` on before-script

File: composer.json (hunk 2 above)
Comment (sstok) [inline composer.json]: And then Symfony 4 may break your future installations :blush: your Composer version constraints should follow semver to have the least change of breakage.

Comment (Ener-Getick) [inline composer.json]: My comment was confusing... I just meant dropping support for symfony 2.1 and 2.2

Comment (sstok) [inline composer.json]: No problem :) :+1: for that.

**Comment by boonkerz** [comment]
any status?

File: Security/Authorization/Interception/MethodSecurityInterceptor.php [hunk 7]
```diff
@@ -50,11 +52,15 @@ class MethodSecurityInterceptor implements MethodInterceptorInterface
     private $runAsManager;
     private $logger;
 
-    public function __construct(SecurityContextInterface $securityContext, AuthenticationManagerInterface $authenticationManager, AccessDecisionManagerInterface $accessDecisionManager,
+    public function __construct($tokenStorage, AuthenticationManagerInterface $authenticationManager, AccessDecisionManagerInterface $accessDecisionManager,
```
Comment (schmittjoh) [inline Security/Authorization/Interception/MethodSecurityInterceptor.php]: Maybe we can add a doc comment what to expect here, i.e. security context or token storage?

File: Twig/SecurityExtension.php (hunk 6 above)
Comment (schmittjoh) [inline Twig/SecurityExtension.php]: @WouterJ was there a reason to not use the interface as suggested by @Ener-Getick and @stof? It seems like the better choice here.

File: Twig/SecurityExtension.php [hunk 8]
```diff
@@ -3,27 +3,33 @@
...
+use Symfony\Component\Security\Core\Authorization\AuthorizationChecker;
 use Symfony\Component\Security\Core\SecurityContextInterface;
 
 class SecurityExtension extends \Twig_Extension
 {
-    private $context;
+    private $authorizationChecker;
 
-    public function __construct(SecurityContextInterface $context)
+    public function __construct($authorizationChecker)
```
Comment (schmittjoh) [inline Twig/SecurityExtension.php]: I would appreciate a doc comment here too which explains what to expect, i.e. security context or authorization checker. Could you add that?

**Comment by stof** [comment]
should this be closed in favor of #202 which builds on top of it ?

**Pull request merged**